### PR TITLE
Implement array dot product

### PR DESCRIPTION
**Pull request by pitrou**
This implements `np.dot`, `np.vdot` and the `@` operator in Python 3.5.  The implementation requires scipy to be installed.

### CODE REVIEW

**Comment by pitrou** [comment]
Hmm, some test failures aren't propagated to the test runner exit code:
https://travis-ci.org/numba/numba/jobs/94460459#L912

**Comment by seibert** [comment]
Can you add an update to the docs in this PR that describes the limitations?  (Only 1 and 2-D arrays, must be contiguous.)

**Comment by pitrou** [comment]
Oh, yes, I forgot to update the docs.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -611,3 +612,336 @@ def scalar_where_impl(cond, x, y):
...
+                            builder.bitcast(out, ll_void_p)))
+    check_blas_return(context, builder, res)
+
+    return builder.load(out)
+
+
+@builtin
+@implement(numpy.dot, types.Kind(types.Array), types.Kind(types.Array))
+@implement('@', types.Kind(types.Array), types.Kind(types.Array))
+def dot_2(context, builder, sig, args):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: docstring?

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -611,3 +612,336 @@ def scalar_where_impl(cond, x, y):
...
+                            data_c, ldc))
+    check_blas_return(context, builder, res)
+
+    return impl_ret_borrowed(context, builder, sig.return_type, out._getvalue())
+
+
+@builtin
+@implement(numpy.dot, types.Kind(types.Array), types.Kind(types.Array),
+           types.Kind(types.Array))
+def dot_3(context, builder, sig, args):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: docstring?

**Comment by pitrou** [comment]
Docstrings added.

**Comment by stuartarchibald** [comment]
I'm broadly happy with this code and implementation of the features. The testing seems to cover common test cases and the implementation on the C side appears to cover most errors that may occur under reasonable circumstances. Perhaps additional functionality of the same class (i.e. BLAS wrapping) could be pulled out into a `_blashelperlib.c` purely to aid navigation of code, not something to do for this PR though. Thanks for the patches, +1 from me.

**Comment by seibert** [comment]
This looks great!

**Pull request merged**